### PR TITLE
Fix GlimmerHMM missing training code

### PR DESCRIPTION
**Pull request by zachcp**
Bioconda requires reviews prior to merging pull-requests (PRs). To facilitate this, once your PR is passing tests and ready to be merged, please add the `please review & merge` label so other members of the bioconda community can have a look at your PR and either make suggestions or merge it. Note that if you are not already a member of the bioconda project (meaning that you can't add this label), please ping `@bioconda/core` so that your PR can be reviewed and merged (please note if you'd like to be added to the bioconda project). Please see https://github.com/bioconda/bioconda-recipes/issues/15332 for more details.

* [ ] I have read the [guidelines for bioconda recipes](https://bioconda.github.io/guidelines.html).
* [ ] This PR adds a new recipe.
* [ ] AFAIK, this recipe **is directly relevant to the biological sciences** (otherwise, please submit to the more general purpose [conda-forge channel](https://conda-forge.org/docs/)).
* [x ] This PR updates an existing recipe.
* [ ] This PR does something else (explain below).

Updating Glimmerhmm for the new build system

### CODE REVIEW

**Comment by zachcp** [comment]
Hi @nextgenusfs ,

I finally got a chance to look at the glimmerhmm module after checking out your PR (https://github.com/bioconda/bioconda-recipes/pull/15613). Can you test that this works on your machine?  It builds on CircleCI for Linux/mac so I think it should be good although I didn't test it on any actual data.

zach cp

File: recipes/glimmerhmm/build.sh [hunk 1]
```diff
@@ -5,16 +5,21 @@ if [ `uname` == Darwin ]; then
...
 mkdir -p $PREFIX/bin
 mkdir -p $PREFIX/GlimmerHMM/trained_dir
 
 # cd into the sources
-cd sources && make
-cd ../train && make
-cd ..
+if [ `uname` == Darwin ]; then
+        cd sources && make CC=$CXX
+        cd ../train && make C=$C CC=$CXX
```
Comment (bgruening) [inline recipes/glimmerhmm/build.sh]: I this really using a C compiler? If so you need to add it in the build-dependencies.

**Comment by nextgenusfs** [comment]
Hi @zachcp and @bgruening.  Okay, after some struggling I got it to build on my Mac -- had to set the Mac SDK in the bootstrap conda_build yaml file.  However, the issue I originally brought up persists, `trainGlimmerHMM` is actually a perl script that is hard coded to look for the binaries and perl modules in its own path, so those need to be moved over to the bin directory as well.  I also found some bugs in the `train/makefile` that was resulting in two binaries not being compiled and therefore failing on OSX, so added a fix for those.  Then also changed shebang line for perl scripts. So here is the `build.sh` that worked for me locally:
```
if [ `uname` == Darwin ]; then
	export MACOSX_DEPLOYMENT_TARGET=10.9
    sed -i.bak "s~#include <malloc.h>~#include <malloc/malloc.h>~g" sources/util.c
    sed -i.bak "s~#include <malloc.h>~#include <malloc/malloc.h>~g" sources/oc1.h
    sed -i.bak "s~#include <malloc.h>~#include <malloc/malloc.h>~g" train/utils.c
fi

#fix typo in makefile in train and make sure make clean wont delete trainGlimmerHMM
sed -i.bak "s~^escoreSTOP2:~scoreSTOP2:~g" train/makefile
sed -i.bak "s~^rfapp:~erfapp:~g" train/makefile

#fix shebang lines in perl scripts
sed -i.bak '1 s|^.*$|#!/usr/bin/env perl|g' train/trainGlimmerHMM
sed -i.bak '1 s|^.*$|#!/usr/bin/env perl|g' bin/glimmhmm.pl

#make directories for storing training data and binaries
mkdir -p $PREFIX/bin
mkdir -p $PREFIX/GlimmerHMM/trained_dir

# cd into the sources
if [ `uname` == Darwin ]; then
    cd sources && make C=$CC CC=$CXX 
    cd ../train && make C=$CC CC=$CXX
    cd ..
else
    cd sources && make CC=$GXX 
    cd ../train && make C=$CC CC=$GXX
    cd ..
fi

#copy the executables and perl modules
cp bin/glimmhmm.pl $PREFIX/bin/.
cp sources/glimmerhmm $PREFIX/bin/.
cp train/trainGlimmerHMM $PREFIX/bin/.
cp train/*.pm $PREFIX/bin/.
cp train/build-icm $PREFIX/bin/.
cp train/build-icm-noframe $PREFIX/bin/.
cp train/build1 $PREFIX/bin/.
cp train/build2 $PREFIX/bin/.
cp train/erfapp $PREFIX/bin/.
cp train/falsecomp $PREFIX/bin/.
cp train/findsites $PREFIX/bin/.
cp train/karlin $PREFIX/bin/.
cp train/score $PREFIX/bin/.
cp train/score2 $PREFIX/bin/.
cp train/scoreATG $PREFIX/bin/.
cp train/scoreATG2 $PREFIX/bin/.
cp train/scoreSTOP $PREFIX/bin/.
cp train/scoreSTOP2 $PREFIX/bin/.
cp train/splicescore $PREFIX/bin/.

# copy the precalculated training data
cp -R trained_dir $PREFIX/GlimmerHMM/

```
Since perl is required for `glimmhmm.pl` and for `trainGlimmerHMM` I'm wondering if perl shouldn't be a runtime dependency? I did at it on my local test as runtime dependency.  So my question is about the perl modules, should those be moved to a conda PerlLib location?  Would need to test if they are then in the @INC or not for this script.  We could also move the train binary dependencies into a sub folder and then modify/patch `trainGlimmerHMM` as there is a place where the script defines the location to the `$script_dir`, this would result in a cleaner install I guess. 

However, if you are both okay with above structure, it does work (I've tested it on actual data).

**Comment by epruesse** [comment]
@biocondabot rebuild

**Comment by zachcp** [comment]
I've updated the build shell as per your suggestions. Lets see if this works.

**Comment by nextgenusfs** [comment]
@BiocondaBot rebuild

**Comment by nextgenusfs** [comment]
The Bioconda-test macOS failed, but I don't understand why.  The package it built is functional on OSX and I don't see the error in the log file.

**Comment by epruesse** [comment]
Looks like the FTP download failed for some reason. Timeout perhaps. I've retriggered the macos build.

File: recipes/glimmerhmm/build.sh [hunk 2]
```diff
@@ -0,0 +1,52 @@
+if [ `uname` == Darwin ]; then
+	export MACOSX_DEPLOYMENT_TARGET=10.9
```
Comment (epruesse) [inline recipes/glimmerhmm/build.sh]: This should be exported by conda-build and not be necessary

File: recipes/glimmerhmm/build.sh [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+cp train/score $PREFIX/bin/.
+cp train/score2 $PREFIX/bin/.
+cp train/scoreATG $PREFIX/bin/.
+cp train/scoreATG2 $PREFIX/bin/.
+cp train/scoreSTOP $PREFIX/bin/.
+cp train/scoreSTOP2 $PREFIX/bin/.
+cp train/splicescore $PREFIX/bin/.
+
+# copy the training data
+cp -R trained_dir $PREFIX/GlimmerHMM/
```
Comment (epruesse) [inline recipes/glimmerhmm/build.sh]: This should not be in the $PREFIX directory. I assume it's data. In that case it should go into $PREFIX/share/glimmerhmm

Comment (zachcp) [reply recipes/glimmerhmm/build.sh]: Good catch. Thanks

**Comment by nextgenusfs** [comment]
Okay, so looks like changing the training dir is fine -- as it is a command line option to `glimmhmm.pl`.  Next query would be about where to put perl modules?  Right now they are being copied over to the ${PREFIX}/bin directory -- as the `trainGlimmerHMM` script looks for them there.  There don't appear to be any other perl modules required for the script to run, so making perl a dependency is not required, although might be best practice to make it a dependency since it is using perl?  Thoughts?

**Comment by epruesse** [comment]
Ugh, yes, I missed that. Copying `score` into `bin/` does not seem to be a good idea. You could look at other recipes. I'm not a perl expert, but there are a bunch that have `Build.PL` files in them that seem to be doing the job. Briefly googling yielded something like `Module::Build::Tiny` that may be available as a package to handle placing the `.pm` files in the right place. Not sure. Most important I think is to keep our namespaces clean (`/bin` as well as the `lib/perl5` folder actually), so that we don't run into issues co-installing other packages with this one.

**Comment by nextgenusfs** [comment]
I've tried a few things, but I'm not sure how to proceed with this.  We need somebody with some Perl experience on how to fix this so that the perl modules located in the source `train` directory are available to the `trainGlimmerHMM` script.

Here is what the header of the `trainGlimmerHMM` looks like:
```
#!/usr/bin/perl

#Copyright (c) 2003  by  Mihaela Pertea.

# no of possible message errors : 179

use strict;
use Getopt::Std;

use Time::localtime;
use File::Basename;
use Cwd;

use FindBin;
use lib $FindBin::Bin;
use orf;
use formtrain;
use dectree_allinfo;
use cfgstat;
use splitiso;

my $workdir=cwd();
my $scriptdir=$FindBin::Bin; # directory where all training programs should be

#print "workdir=$workdir scriptdir=$scriptdir\n";exit;

```
This seems to expect that the perl modules, `orf formtrain dectree_allinfo cfgstat splitiso` are available in the same directory as `trainGlimmerHMM`.  These files in the distribution are located in the `train` directory.  The other thing the script assumes is that several binaries (15 of them) are located in `$scriptdir` -- which also seems to be looking in the same directory as `trainGlimmerHMM`.  Since they don't have very specific names, @epruesse suggested they should be elsewhere.  So I think we could do a sed replacement on the `$scriptdir` and hardcode that to something like `${PREFIX}/share/glimmerhmm/train` -- but there is probably a better way.

**Comment by zachcp** [comment]
@nextgenusfs  thanks for digging in on this one. I think you're spot on - if we change the $SCRIPTDIR we should be able to make everything work.

**Comment by nextgenusfs** [comment]
Okay, so maybe we can just change the `$FindBin::Bin;` to the location of the training directory.  This seems to be the problem, as it calls `lib  $FindBin::Bin;` which adds the script directory to the perl `@INC`, so if we just changed that to the location of the modules should work?  I won't be able to test on my Mac until tonight but I could test a package like above on linux during the day, but a modified build.sh to do could be like this:
```
if [ `uname` == Darwin ]; then
    sed -i.bak "s~#include <malloc.h>~#include <malloc/malloc.h>~g" sources/util.c
    sed -i.bak "s~#include <malloc.h>~#include <malloc/malloc.h>~g" sources/oc1.h
    sed -i.bak "s~#include <malloc.h>~#include <malloc/malloc.h>~g" train/utils.c
fi

#fix typo in makefile in train
sed -i.bak "s~^escoreSTOP2:~scoreSTOP2:~g" train/makefile
sed -i.bak "s~^rfapp:~erfapp:~g" train/makefile

#fix shebang lines in perl scripts and module location
sed -i.bak '1 s|^.*$|#!/usr/bin/env perl|g' train/trainGlimmerHMM
sed -i.bak '1 s|^.*$|#!/usr/bin/env perl|g' bin/glimmhmm.pl
sed -i.bak "s~\$FindBin::Bin;~'${PREFIX}/share/glimmerhmm/train';~g" train/trainGlimmerHMM

#make directories for storing training data and binaries
mkdir -p $PREFIX/bin
mkdir -p $PREFIX/share/glimmerhmm/trained_dir
mkdir -p $PREFIX/share/glimmerhmm/train

# cd into the sources
if [ `uname` == Darwin ]; then
    cd sources && make C=$CC CC=$CXX 
    cd ../train && make C=$CC CC=$CXX
    cd ..
else
    cd sources && make CC=$GXX 
    cd ../train && make C=$CC CC=$GXX
    cd ..
fi

# copy the executables and perl modules to bin
cp bin/glimmhmm.pl $PREFIX/bin/.
cp sources/glimmerhmm $PREFIX/bin/.
cp train/trainGlimmerHMM $PREFIX/bin/.

# copy binaries and training perl modules to sub folder
cp train/*.pm $PREFIX/share/glimmerhmm/train/.
cp train/build-icm $PREFIX/share/glimmerhmm/train/.
cp train/build-icm-noframe $PREFIX/share/glimmerhmm/train/.
cp train/build1 $PREFIX/share/glimmerhmm/train/.
cp train/build2$PREFIX/share/glimmerhmm/train/.
cp train/erfapp $PREFIX/share/glimmerhmm/train/.
cp train/falsecomp $PREFIX/share/glimmerhmm/train/.
cp train/findsites $PREFIX/share/glimmerhmm/train/.
cp train/karlin $PREFIX/share/glimmerhmm/train/.
cp train/score $PREFIX/share/glimmerhmm/train/.
cp train/score2 $PREFIX/share/glimmerhmm/train/.
cp train/scoreATG $PREFIX/share/glimmerhmm/train/.
cp train/scoreATG2 $PREFIX/share/glimmerhmm/train/.
cp train/scoreSTOP $PREFIX/share/glimmerhmm/train/.
cp train/scoreSTOP2$PREFIX/share/glimmerhmm/train/.
cp train/splicescore$PREFIX/share/glimmerhmm/train/.

# copy the precalculated training data
cp -R trained_dir $PREFIX/share/glimmerhmm/train/
```
And then I think add perl to host and run in yaml.
```
package:
  name: glimmerhmm
  version: "3.0.4"

source:
  md5: af3626a5eceb2893ba4e19373dfad1b0
  url: ftp://ccb.jhu.edu/pub/software/glimmerhmm/GlimmerHMM-3.0.4.tar.gz

build:
  number: 3

requirements:
  build:
    - {{ compiler('c') }}
    - {{ compiler('cxx') }}
  host:
    - perl
  run:
    - perl

test:
  commands:
    - glimmerhmm -h > /dev/null

about:
  home: https://ccb.jhu.edu/software/glimmerhmm/
  license: Artistic License
  summary: "GlimmerHMM is a gene finder based on a Generalized Hidden Markov Model (GHMM)"
```

**Comment by epruesse** [comment]
ping @bioconda/perl - could you have a look?

**Comment by nextgenusfs** [comment]
Hi @zachcp and @epruesse. A few more tests and slightly different strategy than I posted above, this is now working on my Mac and should patch the perl script.  I found a few more errors in the `train/makefile` that weren't compiling some of the programs (thus the precompiled binaries were getting moved over and they obviously failed). So I setup a `make clean` to ensure that compiled binaries are being generated.  What I did with the perl modules and library location in the `trainGlimmerHMM` perl script was to borrow what was being done in Jbrowse to a location of perl modules, i.e. see https://github.com/GMOD/jbrowse/blob/master/bin/prepare-refseqs.pl.

The binaries and perl modules for `trainGlimmerHMM` are being installed in `${PREFIX}/share/glimmerhmm/train`.

I validated the install is working by building with `bioconda-utils build --packages=glimmerhmm --force --keep-image --build-image recipes/ config.yml` and then extracting the image and testing it.

Use same `meta.yaml` as above and then working `build.sh` is here:
```
if [ `uname` == Darwin ]; then
    sed -i.bak "s~#include <malloc.h>~#include <malloc/malloc.h>~g" sources/util.c
    sed -i.bak "s~#include <malloc.h>~#include <malloc/malloc.h>~g" sources/oc1.h
    sed -i.bak "s~#include <malloc.h>~#include <malloc/malloc.h>~g" train/utils.c
fi

# fix typos in train makefile
sed -i.bak "s|^escoreSTOP2:|scoreSTOP2:|g" train/makefile
sed -i.bak "s|^rfapp:|erfapp:|g" train/makefile
sed -i.bak "s| trainGlimmerHMM||g" train/makefile
sed -i.bak "s|all:    build-icm|all:    misc.o build-icm.o build-icm-noframe.o build-icm|g" train/makefile

# fix perl scripts
sed -i.bak '1 s|^.*$|#!/usr/bin/env perl|g' train/trainGlimmerHMM
sed -i.bak 's|FindBin;|FindBin qw($RealBin);|g' train/trainGlimmerHMM
sed -i.bak 's|$FindBin::Bin;|"$RealBin/../share/glimmerhmm/train";|g' train/trainGlimmerHMM
sed -i.bak '1 s|^.*$|#!/usr/bin/env perl|g' bin/glimmhmm.pl

# make directories for storing training data and binaries
mkdir -p $PREFIX/bin
mkdir -p $PREFIX/share/glimmerhmm
mkdir -p $PREFIX/share/glimmerhmm/train

# cd into the sources
if [ `uname` == Darwin ]; then
        cd sources && make CC=$CXX
        cd ../train && make clean && make all C=$CC CC=$CXX
        cd ..
else
        cd sources && make CC=$GXX 
        cd ../train && make clean && make all C=$CC CC=$CXX
        cd ..
fi

# copy the executables	
cp bin/glimmhmm.pl $PREFIX/bin/.
cp sources/glimmerhmm $PREFIX/bin/.
cp train/trainGlimmerHMM $PREFIX/bin/.
cp train/build-icm $PREFIX/share/glimmerhmm/train/.
cp train/build-icm-noframe $PREFIX/share/glimmerhmm/train/.
cp train/build1 $PREFIX/share/glimmerhmm/train/.
cp train/build2 $PREFIX/share/glimmerhmm/train/.
cp train/erfapp $PREFIX/share/glimmerhmm/train/.
cp train/falsecomp $PREFIX/share/glimmerhmm/train/.
cp train/findsites $PREFIX/share/glimmerhmm/train/.
cp train/karlin $PREFIX/share/glimmerhmm/train/.
cp train/score $PREFIX/share/glimmerhmm/train/.
cp train/score2 $PREFIX/share/glimmerhmm/train/.
cp train/scoreATG $PREFIX/share/glimmerhmm/train/.
cp train/scoreATG2 $PREFIX/share/glimmerhmm/train/.
cp train/scoreSTOP $PREFIX/share/glimmerhmm/train/.
cp train/scoreSTOP2 $PREFIX/share/glimmerhmm/train/.
cp train/splicescore $PREFIX/share/glimmerhmm/train/.

# copy the perl modules
cp train/*.pm $PREFIX/share/glimmerhmm/train/.

# copy the training data
cp -R trained_dir $PREFIX/share/glimmerhmm/
```

**Comment by zachcp** [comment]
Nice work, @nextgenusfs.  I'm traveling this week. Any chance you can apply these changes to this branch? I am happy to review them and get this live.

**Comment by nextgenusfs** [comment]
Sure @zachcp -- just pushed the changes. @epruesse not sure if you have to approve the changes or not, but can you restart the build?

File: recipes/glimmerhmm/build.sh [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+mkdir -p $PREFIX/share/glimmerhmm
+mkdir -p $PREFIX/share/glimmerhmm/train
+
+# cd into the sources
+if [ `uname` == Darwin ]; then
+        cd sources && make CC=$CXX
+        cd ../train && make clean && make all C=$CC CC=$CXX
+        cd ..
+else
+        cd sources && make CC=$GXX
```
Comment (epruesse) [inline recipes/glimmerhmm/build.sh]: Replace GXX with CXX

File: recipes/glimmerhmm/build.sh [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+# cd into the sources
+if [ `uname` == Darwin ]; then
+        cd sources && make CC=$CXX
+        cd ../train && make clean && make all C=$CC CC=$CXX
+        cd ..
+else
+        cd sources && make CC=$GXX 
+        cd ../train && make clean && make all C=$CC CC=$CXX
+        cd ..
+fi
```
Comment (epruesse) [inline recipes/glimmerhmm/build.sh]: Actually, this should be the same thing for osx and linux. Why the switch? CXX is set to the right thing on either architecture.

File: recipes/glimmerhmm/build.sh [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+cp train/falsecomp $PREFIX/share/glimmerhmm/train/.
+cp train/findsites $PREFIX/share/glimmerhmm/train/.
+cp train/karlin $PREFIX/share/glimmerhmm/train/.
+cp train/score $PREFIX/share/glimmerhmm/train/.
+cp train/score2 $PREFIX/share/glimmerhmm/train/.
+cp train/scoreATG $PREFIX/share/glimmerhmm/train/.
+cp train/scoreATG2 $PREFIX/share/glimmerhmm/train/.
+cp train/scoreSTOP $PREFIX/share/glimmerhmm/train/.
+cp train/scoreSTOP2 $PREFIX/share/glimmerhmm/train/.
+cp train/splicescore $PREFIX/share/glimmerhmm/train/.
```
Comment (epruesse) [inline recipes/glimmerhmm/build.sh]: Why the `/.` at the end of the dirnames?

File: recipes/glimmerhmm/build.sh (hunk 4 above)
Comment (epruesse) [reply recipes/glimmerhmm/build.sh]: Actually, might be able to call
```
make -C sources CC=$CXX
make -C train C=$CC CC=$CXX
```

What's with CFLAGS/CXXFLAGS/LDFLAGS etc? Are they used by the makefile or do they need to be added?

Conda has some special flags we should try to use if possible. They are exported just as $CC is.

File: recipes/glimmerhmm/build.sh (hunk 6 above)
Comment (nextgenusfs) [reply recipes/glimmerhmm/build.sh]: That's just how @zachcp had the original one. will edit.

File: recipes/glimmerhmm/build.sh (hunk 4 above)
Comment (nextgenusfs) [reply recipes/glimmerhmm/build.sh]: I don't know about the CFLAGS etc -- was just trying to get it to compile!

**Comment by nextgenusfs** [comment]
@epruesse I'm not sure how the new review/approval system works.  But in order for this to build again does it require you to "approve" your requested changes?  Seems strange that it hasn't tried to rebuild for ~ 6 days.

**Comment by epruesse** [comment]
@Biocondabot rebuild

Elmar

--
message sent from mobile device

**Comment by epruesse** [comment]
@nextgenusfs That wasn't the Bot, it was CircleCI not building the newest push. That sometimes happens. It just misses the pull_request/synchronize event. The bot command `rebuild` triggers a rebuild "manually".

**Comment by epruesse** [comment]
@biocondabot merge

**Comment by epruesse** [comment]
@BiocondaBot merge

**Comment by zachcp** [comment]
I'm back and it looks like you have all been busy this week. Thank you both.   I will merge now.

**Pull request merged**